### PR TITLE
REST high-level client: add support for Indices Update Settings API [take 2]

### PR DESCRIPTION
**Pull request by olcbean**
Relates to #27205

### CODE REVIEW

**Comment by javanna** [comment]
hi @olcbean thanks for your effort, of trying to address the issue caused in x-pack tests that you can't even look at :)

I didn't see this coming unfortunately. The problem in your previous PR, and partially in this updated one too, is that we change equals/hashcode implementation for base classes that are subclassed by a lot of classes. At that point we should also update all those classes otherwise e.g. two different cluster health requests are deemed equal as we only use fields from its base class to do the comparison. Rather than making this PR huge though I think I would only modify equals/hashcode for the request that we are working on without affecting its base classes hence all of the sublcasses of such base classes. One day, or in a follow-up PR, we can do this properly, but for now I think that makes the change too big and unnecessary. How does this sound to you?

**Comment by olcbean** [comment]
@javanna it is definitely better to address the equals / hashCode gradually rather than as part of this PR ;)

**Comment by olcbean** [comment]
btw is this the first Request for which `AbstractStreamableTestCase` and `AbstractXContentTestCase` were added ? Do you think such tests should be added for the majority of the Requests ( while implementing the high level client ) ?

**Comment by javanna** [comment]
> btw is this the first Request for which AbstractStreamableTestCase and AbstractXContentTestCase were added ? Do you think such tests should be added for the majority of the Requests ( while implementing the high level client ) ?

Yes it may be the first one, most of the request tests are probably just `ESTestCase`. As we've seen in this PR, for requests it's a bit more complicated than for responses, because requests are made of url parameters and optional body, so there is a mismatch between transport serialization (body plus parameters) and xcontent serialization (body only) which makes things like equality comparisons and assertions complicated. I would go case by case. Probably dividing the two tests is the best solution in most cases, but we will see as we go I guess.

File: server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/UpdateSettingsRequest.java [hunk 1]
```diff
@@ -239,19 +239,22 @@ public String toString() {
...
         }
-        return false;
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        UpdateSettingsRequest that = (UpdateSettingsRequest) o;
+        return Objects.equals(settings, that.settings)
+                && Objects.equals(indicesOptions, that.indicesOptions)
+                && Objects.equals(preserveExisting, that.preserveExisting)
+                && Arrays.equals(indices, that.indices);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/UpdateSettingsRequest.java]: don't you need to take also timeout and masterNodeTimeout into account?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/UpdateSettingsRequest.java]: I would suspect proper equals/hashcode tests to fail currently, maybe also tests need to be updated to mutate such fields.

Comment (olcbean) [reply server/src/main/java/org/elasticsearch/action/admin/indices/settings/put/UpdateSettingsRequest.java]: Now the equals and the hashCode use all request fields except for the flatSettings ( as the flatSettings are not serialized ).

**Comment by javanna** [comment]
test this please

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
retest this please

**Review by javanna — APPROVED** [review]

**Comment by javanna** [comment]
Thanks @olcbean once again and sorry it took me a while to merge it.

**Pull request merged**